base/compiler/ssair/inlining.jl

```diff
@@ -1355,16 +1356,18 @@ function compute_inlining_cases(@nospecialize(info::CallInfo), flag::UInt32, sig
             joint_effects = merge_effects(joint_effects, info_effects(result, match, state))
             split_fully_covered |= match.fully_covers
             if !validate_sparams(match.sparams)
-                if !match.fully_covers
-                    handled_all_cases = false
-                    continue
-                end
-                if revisit_idx === nothing
-                    revisit_idx = (i, j, all_result_count)
+                if match.fully_covers
+                    if revisit_idx === nothing
+                        revisit_idx = (i, j, all_result_count)
+                    else
+                        handled_all_cases = false
+                        revisit_idx = nothing
+                    end
                 else
                     handled_all_cases = false
-                    revisit_idx = nothing
                 end
+            elseif !(match.spec_types <: match.method.sig) # the requirement for correct union-split
+                handled_all_cases = false
             else
                 handled_all_cases &= handle_any_const_result!(cases,
                     result, match, argtypes, info, flag, state; allow_abstract=true, allow_typevars=false)
```

Comment:
```suggestion
                    result, match, argtypes, info, flag, state; allow_abstract=handled_all_cases, allow_typevars=false)
```
This can only permit later abstract calls if we proved that all previous calls were disjoint from this. One conservative way to express that requirement is to only allow isdispatchtuple here.

Reply:
Isn't this filtering enough for this?
https://github.com/JuliaLang/julia/blob/63e75e899997d41058ef70ca687b4e57732d72c3/base/compiler/ssair/inlining.jl#L1391-L1393

Reply:
Additionally, if `nunion != 1`, then the list of matches cannot be assumed to be sorted, so we can only allow the abstract calls that we can prove are disjoint from the other union elements, since we do not attempt to merge-sort them to combine them (the merge sort might fail) but we simply concat the lists.
```suggestion
                    result, match, argtypes, info, flag, state; allow_abstract=handled_all_cases && nunion == 1, allow_typevars=false)
```

Reply:
Ah, okay, I wasn't expecting we would over-estimate this then filter it later. That late filter is actually a bit more strict than necessary then, since it doesn't consider the ordering. But that does at least seem valid.